File: app/src/main/java/de/westnordost/streetcomplete/quests/building_levels/AddBuildingLevelsForm.java
package de.westnordost.streetcomplete.quests.building_levels;

import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import javax.inject.Inject;

import de.westnordost.streetcomplete.Injector;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment;
import de.westnordost.streetcomplete.quests.LastPickedValuesStore;
import de.westnordost.streetcomplete.util.TextChangedWatcher;

public class AddBuildingLevelsForm extends AbstractQuestFormAnswerFragment
{
	public static final String BUILDING_LEVELS = "building_levels";
	public static final String ROOF_LEVELS = "roof_levels";

	private static final int MAX_FAVS = 1;

	private EditText levelsInput, roofLevelsInput;

	@Inject LastPickedValuesStore favs;

	@Override public void onCreate(Bundle inState)
	{
		super.onCreate(inState);
		Injector.instance.getApplicationComponent().inject(this);
	}

	@Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
									   Bundle savedInstanceState)
	{
		View view = super.onCreateView(inflater, container, savedInstanceState);

		View contentView = setContentView(R.layout.quest_building_levels);

		TextWatcher onTextChangedListener = new TextChangedWatcher(this::checkIsFormComplete);

		levelsInput = contentView.findViewById(R.id.levelsInput);
		levelsInput.requestFocus();
		levelsInput.addTextChangedListener(onTextChangedListener);
		roofLevelsInput = contentView.findViewById(R.id.roofLevelsInput);
		roofLevelsInput.addTextChangedListener(onTextChangedListener);

		addOtherAnswer(R.string.quest_buildingLevels_answer_multipleLevels, () ->
		{
			new AlertDialog.Builder(getActivity())
					.setMessage(R.string.quest_buildingLevels_answer_description)
					.setPositiveButton(android.R.string.ok, null)
					.show();
		});

		View btnPickLast = contentView.findViewById(R.id.btn_pickLast);
		LinkedList<String> lastPicked = favs.getLastPicked(getClass().getSimpleName());
		if(lastPicked.isEmpty())
		{
			btnPickLast.setVisibility(View.GONE);
		}
		else
		{
			btnPickLast.setVisibility(View.VISIBLE);

			String[] favValues = lastPicked.getFirst().split("#");

			TextView lastLevelsText = contentView.findViewById(R.id.lastLevelsText);
			lastLevelsText.setText(favValues[0]);

			TextView lastRoofLevelsText = contentView.findViewById(R.id.lastRoofLevelsText);
			lastRoofLevelsText.setText(favValues.length > 1 ? favValues[1] : " ");

			btnPickLast.setOnClickListener(v -> {
				levelsInput.setText(lastLevelsText.getText());
				roofLevelsInput.setText(lastRoofLevelsText.getText());
				btnPickLast.setVisibility(View.GONE);
			});
		}

		return view;
	}

	@Override protected void onClickOk()
	{
		List<Integer> favValues = new LinkedList<>();
		Bundle answer = new Bundle();

		int buildingLevels = Integer.parseInt(getLevels());
		answer.putInt(BUILDING_LEVELS, buildingLevels);
		favValues.add(buildingLevels);

		if(!getRoofLevels().isEmpty())
		{
			int roofLevels = Integer.parseInt(getRoofLevels());
			answer.putInt(ROOF_LEVELS, roofLevels);
			favValues.add(roofLevels);
		}
		favs.addLastPicked(getClass().getSimpleName(), TextUtils.join("#", favValues), MAX_FAVS);
		applyAnswer(answer);
	}

	private String getLevels() { return levelsInput.getText().toString().trim(); }
	private String getRoofLevels() { return roofLevelsInput.getText().toString().trim(); }

	@Override public boolean isFormComplete() { return !getLevels().isEmpty(); }
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/building_levels/AddBuildingLevelsForm.kt
package de.westnordost.streetcomplete.quests.building_levels

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.preferences.Preferences
import de.westnordost.streetcomplete.databinding.QuestBuildingLevelsBinding
import de.westnordost.streetcomplete.quests.AbstractOsmQuestForm
import de.westnordost.streetcomplete.quests.AnswerItem
import de.westnordost.streetcomplete.ui.util.content
import de.westnordost.streetcomplete.util.takeFavourites
import org.koin.android.ext.android.inject

class AddBuildingLevelsForm : AbstractOsmQuestForm<BuildingLevels>() {

    override val contentLayoutResId = R.layout.quest_building_levels
    private val binding by contentViewBinding(QuestBuildingLevelsBinding::bind)

    private val prefs: Preferences by inject()
    private lateinit var levels: MutableState<String>
    private lateinit var roofLevels: MutableState<String>
    override val otherAnswers = listOf(
        AnswerItem(R.string.quest_buildingLevels_answer_multipleLevels) { showMultipleLevelsHint() }
    )

    private val lastPickedAnswers by lazy {
        prefs.getLastPicked(this::class.simpleName!!)
            .map { value ->
                value.split("#")
                    .let { BuildingLevels(it[0].toInt(), it.getOrNull(1)?.toInt()) }
            }
            .takeFavourites(n = 5, history = 15, first = 1)
            .sortedWith(compareBy<BuildingLevels> { it.levels }.thenBy { it.roofLevels })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.questBuildingLevelsBase.content {
            levels = rememberSaveable { mutableStateOf(element.tags["building:levels"] ?: "") }
            roofLevels = rememberSaveable { mutableStateOf(element.tags["roof:levels"] ?: "") }
            BuildingLevelsForm(
                levels = levels.value,
                onLevelsChange = {
                    levels.value = it
                    checkIsFormComplete()
                },
                roofLevels = roofLevels.value,
                onRoofLevelsChange = {
                    roofLevels.value = it
                    checkIsFormComplete()
                },
                previousBuildingLevels = lastPickedAnswers
            )
        }
    }

    override fun onClickOk() {
        val answer = BuildingLevels(
            levels.value.toInt(),
            roofLevels.value.takeIf { it.isNotEmpty() }?.toInt()
        )
        prefs.addLastPicked(
            this::class.simpleName!!,
            listOfNotNull(answer.levels, answer.roofLevels).joinToString("#")
        )
        applyAnswer(answer)
    }

    private fun showMultipleLevelsHint() {
        activity?.let {
            AlertDialog.Builder(it)
                .setMessage(R.string.quest_buildingLevels_answer_description)
                .setPositiveButton(android.R.string.ok, null)
                .show()
        }
    }

    override fun isFormComplete(): Boolean {
        val roofShape = element.tags["roof:shape"]
        val hasNonFlatRoofShape = roofShape != null && roofShape != "flat"
        val roofLevelsAreOptional = countryInfo.roofsAreUsuallyFlat && !hasNonFlatRoofShape

        return levels.value.isValidLevel()
            && (
                roofLevelsAreOptional && roofLevels.value.isEmpty()
                || roofLevels.value.isValidLevel()
            )
    }

    private fun String.isValidLevel(): Boolean =
        toIntOrNull()?.takeIf { it >= 0 } != null
}
